### ISSUE TITLE
[@metadata][index] bug(s)?

### ISSUE THREAD

**Issue by erik-stephens** [logstash-plugins/logstash-input-beats#6]
The attached configuration was run on OS X with following versions.  I'm seeing same behavior on x86_64 as well.

- elasticsearch 1.7.3
- logstash 1.5.4 with latest beats input plugin
- filebeat 1.0.0-beta4

I see two issues or potential sources of confusion:

1. The beats plugin seems to reserve `[@metadata][index]`: I can't remove it or redefine it.
2. The interpolated index does not use the event's timestamp but the shipped timestamp instead.

Regarding item 1, I can workaround by using something other than `[@metadata][index]` in my config.  I recommend addressing in the docs so others don't hit this unexpected name clash.  Regarding item 2, not as important in my use case since I define the index explicitly but I can imagine that impacting quite a few users.

Hopefully this is enough information to reproduce:
```
$ ~/Downloads/logstash-1.5.4/bin/logstash -f /tmp/logstash.conf &
Oct 25, 2015 9:25:51 AM org.elasticsearch.plugins.PluginsService <init>
INFO: [logstash-ErikStephens.local-90663-11760] loaded [], sites []
Logstash startup completed

$ rm -f .filebeat; filebeat -e -d publish -c filebeat.yaml
publish.go:185: DBG  create output worker: 0x0, 0x0
publish.go:222: WARN No output is defined to store the topology. The server fields might not be filled.
async.go:95: DBG  create bulk processing worker (interval=1s, bulk size=10000)
crawler.go:77: INFO All prospectors initialised with 0 states to persist
output.go:37: DBG  output worker: no events to publish
output.go:37: DBG  output worker: no events to publish
client.go:58: DBG  send event
preprocess.go:37: DBG  preprocessor
publish.go:87: DBG  Publish: {
  "count": 1,
  "fields": null,
  "fileinfo": {},
  "line": 1,
  "message": "HELLO",
  "offset": 0,
  "shipper": "test",
  "source": "test.log",
  "timestamp": "2015-10-25T16:26:10.413Z",
  "type": "log"
}
preprocess.go:91: DBG  preprocessor forward
output.go:42: DBG  output worker: publish 1 events
output.go:37: DBG  output worker: no events to publish

{
       "message" => "HELLO",
      "@version" => "1",
    "@timestamp" => "2015-01-01T12:00:00.000Z",
          "type" => "log",
         "count" => 1,
        "fields" => nil,
      "fileinfo" => {},
          "line" => 1,
        "offset" => 0,
       "shipper" => "test",
        "source" => "test.log",
     "timestamp" => "2015-10-25T16:26:10.413Z",
            "ts" => "2015-01-01T12:00:00Z",
     "@metadata" => {
        "index" => "filebeat-2015.10.25",
         "type" => "log"
    }
}

$ curl localhost:9200/_cat/indices?v
health status index               pri rep docs.count docs.deleted store.size pri.store.size
yellow open   filebeat-2015.10.25   5   1          1            0        5kb            5kb
```

cat /tmp/logstash.conf
```
input {
  stdin {
    type => stdin
  }
  beats {
    type => beats
    port => 12345
  }
}
filter {
  mutate {
    replace => [
      'ts', '2015-01-01T12:00:00Z'
    ]
    remove_field => [
      '[@metadata][index]'
    ]
  }
  date {
    match => [ 'ts', 'ISO8601' ]
  }
}
output {
  stdout {
    codec => rubydebug { metadata => true }
  }
  elasticsearch {
    host => [ 'localhost' ]
    protocol => 'transport'
    cluster => 'elasticsearch'
    action => update
    doc_as_upsert => true
    index => '%{[@metadata][index]}'
    document_id => 'foo'
  }
}
```

**Comment by ruflin** [comment]
@erik-stephens This also relates to: https://github.com/elastic/filebeat/issues/120

**Comment by urso** [comment]
Checking logstash source processing [@metadata] via mutate plugin might be limited. Looks like 'update', 'replace' or 'merge' might be able to update [@metadata][index]. But 'remove_field' seems unable to access @metadata.

**Comment by ph** [comment]
@erik-stephens thanks for reporting this.

@urso @ruflin Not sure what would be the best behavior here, its there any specific reason why the shipper is sending the index? Because without making sense of the actual data it will only be related to the current read date?

**Comment by jordansissel** [comment]
> But 'remove_field' seems unable to access @metadata.

This may be a bug, but can you tell me why you want to delete something from metadata?

**Comment by erik-stephens** [comment]
My real configuration has this near the end in to define a default index:
```
if ! [@metadata][index] {
  mutate {
    replace => [ '[@metadata][index]', 'logstash-%{+YYYY.MM.dd}' ]
  }
}
```
That's why I wanted to remove/unset.  Thanks @urso, good to know.  I mis-spoke: it can be redefined.  I thought I tried without success.  I think I tried to replace it with `null` which I just learned is really the string 'null'.

I couldn't find in the beats input source where [@metadata][index] was being defined.  If the shipper is sending it, then I would expect it to show in the `-d publish` output.

So maybe not bug(s) and better addressed in the docs?

- Have the filter `remove_field` docs note that [@metadata] fields are ignored.
- Have the beats plugin or filebeat docs provide more detail about how the index is defined.

**Comment by urso** [comment]
behavior change in PR elastic/libbeat#231

We will send [@metadata][beat] for users having better control over indexing in ES.

**Comment by urso** [comment]
@erik-stephens is change to [@metadata][beat] good for you. I'd count this report as problem in beats, not input plugin. I think with changes in libbeat this one can be closed

**Comment by erik-stephens** [comment]
LGTM, particularly the changes in logstash.go.  Regarding [@metadata][index] vs [@metadata][beat], I think the main source of my confusion was not knowing about the libbeat docs.  Between filebeat, logstash beats input, packetbeat, and libbeat, I suspect others might get confused as well.  Or worse, the docs may diverge.  Or, maybe just because I'm new to the beats world.  Anyway, thanks for addressing!

Regarding `remove_field` not able to operate on [@metadata], is that on anyone's radar?  Should I open a separate issue, and if so, where?  logstash-core?  I don't think it necessarily a bug, but I think it should be noted in the docs somewhere.

**Comment by jordansissel** [comment]
@erik-stephens what is the purpose of remove_field on `@metadata`? For data that isn't included in any output, what's the goal of removing it?

**Comment by jordansissel** [comment]
@erik-stephens but yes, opening a bug probably on the main logstash repo (elastic/logstash) since this limitation/problem occurs in that code base and affects all/most things ;P

**Comment by erik-stephens** [comment]
@jordansissel I think the use case is to be able to override/undo anything that something upstream might set, especially as @metadata is used more & more.  In my specific case, adding a beats input conflicted with my existing configuration, which is fine, but I would still like a way for my configuration to take precedence.  Maybe `remove_field` isn't the best way but was the only option I could think of at the time.

**Comment by jordansissel** [comment]
ahh, fair. Yeah let's get a bug filed about it and we can fix it. Thanks for the background on why you need it fixed! :)

**Comment by JacqueCC** [comment]
How can I pass temporary data from the filter to the index field of the elasticsearch output in logstash?

I did like this:

    filter {
      mutate { add_field => { "[@metadata][TEMP_DATA]" => "%{custom_index}" } }
      mutate { remove_field => ["custom_index"] }
    }
    
    output {
      elasticsearch {
        index => "%{[@metadata][TEMP_DATA]}"
      }
    }

But it did not work. It also did not create an index %{[@metadata][TEMP_DATA]} if someone wonders, but it sent data to elasticsearch. I can see that because I am using ELK 5 and there is a Monitoring page in Kibana where I see the Logstash Node. When I click on it I can see the data that the Logstash Node has received.

I have created indices where the data should be saved in elasticsearch, but the data is been delivered somewhere also I think.

Does anyone have an idea how I could solve that problem?
